src/api/libcellml/component.h

```diff
@@ -0,0 +1,63 @@
+#ifndef LIBCELLML_LIBCELLML_COMPONENT_H
+#define LIBCELLML_LIBCELLML_COMPONENT_H
+
+#include <string>
+#include <memory>
+
+#include "libcellml/libcellml_export.h"
+#include "libcellml/nameable.h"
+
+namespace libcellml {
+
+/**
+ * @brief The Model class.
+ * The Model class is for representing a CellML Model.
+ */
+class LIBCELLML_EXPORT Component: public Nameable
+{
+public:
+    /**
+     * @brief Component Constructor.
+     * Default Component constructor.
+     */
+    Component();
+    /**
+     * @brief Component Destructor.
+     * Default Component destructor.
+     */
+    ~Component();
+
+    /**
+     * @brief serialise the Model.
+     * This method serialises the Model to a std:string representation in the
+     * desired format.
+     * @sa deserialise
+     * @param format The format to serialise the Model to.
+     * @return std::string representation of the Model in the given format.
+     */
+    std::string serialise(CELLML_FORMATS format) const;
+
+    Component(Component&& rhs); /**< move constructor */
+    Component& operator=(Component&& rhs); /**< move assignment */
+
+    /**
+     * @brief add a child component to this component.
+     * Add a child component to this component.
+     * @param c the component to add.
+     */
+    void addComponent(Component &c);
+
+private:
+
+    Component(const Component&) = delete; /**< non-copyable */
+    Component& operator=(const Component&) = delete; /**< non-copyable */
+
+    struct ComponentImpl; /**< Forward declaration for pImpl idiom. */
+    std::unique_ptr<ComponentImpl> mPimpl; /**< Private member to implementation pointer */
```

Comment:
I was thinking this whole class should have been a handle to a component in which case I expected a shared_ptr here. On thinking about it, it is acceptable as long as the client only ever accesses it via a std::shared_ptr<libcellml::Component>. In that case I would expect the constructors to be private and instead only publicly expose a static create() method which returns a std::shared_ptr, meaning whether you are creating a new component or getting a component from a model or other component, you are always dealing with a shared_ptr to it. Similarly all addComponent methods would take the shared_ptr reference too.
The alternative to a static create() method on the class is to require a model to create a component via a Model::createComponent() method which supports a system where the Model knows all Components under it at all times and permits global defaults to be passed to the constructor, if that's useful.